### PR TITLE
[SPARK-48518][CORE] Make LZF compression be able to run in parallel

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR introduced a config that turns on LZF compression to parallel mode via using PLZFOutputStream.

FYI, https://github.com/ning/compress?tab=readme-ov-file#parallel-processing

### Why are the changes needed?

Improve performance

```
[info] OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
[info] Apple M2 Max
[info] Compress large objects:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] -----------------------------------------------------------------------------------------------------------------------------
[info] Compression 1024 array values in 7 threads                12             13           1          0.1       11788.2       1.0X
[info] Compression 1024 array values single-threaded             23             23           0          0.0       22512.7       0.5X
```

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

benchmark
### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: core/benchmarks/LZFBenchmark-jdk21-results.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+Compress small objects:                           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+--------------------------------------------------------------------------------------------------------------------------------
+Compression 256000000 int values in parallel                598            600           2        428.2           2.3       1.0X
+Compression 256000000 int values single-threaded            568            570           2        451.0           2.2       1.1X
+
+OpenJDK 64-Bit Server VM 21.0.3+9-LTS on Linux 6.5.0-1021-azure
+AMD EPYC 7763 64-Core Processor
+Compress large objects:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+-----------------------------------------------------------------------------------------------------------------------------
+Compression 1024 array values in 1 threads                39             45           5          0.0       38475.4       1.0X
```
Comment (yaooqinn) [inline core/benchmarks/LZFBenchmark-jdk21-results.txt:16]: With GitHub standard action runners, it seems that we only get 1 thread.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -2031,6 +2031,13 @@ package object config {
       .intConf
       .createWithDefault(1)
 
+  private[spark] val IO_COMPRESSION_LZF_PARALLEL =
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2034]: Should we add an explanation of this configuration in `configuration.md`

File: core/benchmarks/LZFBenchmark-jdk21-results.txt (hunk 1 above)
Comment (LuciferYang) [reply core/benchmarks/LZFBenchmark-jdk21-results.txt:16]: hmm... 

```
[info] OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.5
[info] Apple M2 Max
[info] Compress small objects:                           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] --------------------------------------------------------------------------------------------------------------------------------
[info] Compression 256000000 int values in parallel                548            550           2        467.0           2.1       1.0X
[info] Compression 256000000 int values single-threaded            522            523           1        490.5           2.0       1.1X
[info] Running benchmark: Compress large objects
[info]   Running case: Compression 1024 array values in 8 threads
[info]   Stopped after 123 iterations, 2009 ms
[info]   Running case: Compression 1024 array values single-threaded
[info]   Stopped after 83 iterations, 2003 ms
[info] OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.5
[info] Apple M2 Max
[info] Compress large objects:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] -----------------------------------------------------------------------------------------------------------------------------
[info] Compression 1024 array values in 8 threads                12             16          13          0.1       11546.1       1.0X
[info] Compression 1024 array values single-threaded             23             24           1          0.0       22767.9       0.5X

```

I ran this benchmark locally, and it seems that the performance of `Compression 256000000 int values single-threaded` and `Compression 256000000 int values in parallel` is almost the same

Comment (yaooqinn) [reply core/benchmarks/LZFBenchmark-jdk21-results.txt:16]: I guess the rate is limited by the producer

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 2 above)
Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2034]: thanks, addressed

**Review by LuciferYang — APPROVED** [review]
LGTM

**Comment by yaooqinn** [comment]
Thank you @LuciferYang

Merged to master

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 3]
```diff
@@ -170,9 +171,14 @@ class LZ4CompressionCodec(conf: SparkConf) extends CompressionCodec {
  */
 @DeveloperApi
 class LZFCompressionCodec(conf: SparkConf) extends CompressionCodec {
+  private val parallelCompression = conf.get(IO_COMPRESSION_LZF_PARALLEL)
 
   override def compressedOutputStream(s: OutputStream): OutputStream = {
-    new LZFOutputStream(s).setFinishBlockOnFlush(true)
+    if (parallelCompression) {
+      new PLZFOutputStream(s)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala:178]: This is creating a threadpool per `compressedOutputStream` - which can end up being quite expensive (num thread is num processors + some 'interesting' logic which tries to modulate it) .

Did you get a chance to try this on some nontrivial jobs ? Very curious about the experience.
Given this is turned off by default, dont see any concerns with the change itself though ! Would be a good way to understand the impact.

**Pull request closed without merging**